### PR TITLE
Document how we've used domain driven design

### PR DESCRIPTION
**Pull request by junlarsen**
Quite a bit of yapping but it should give beginners an idea of why we do it, how it's useful, and how we implement it

### CODE REVIEW

File: docs/domain-driven-design.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
+# Domain Driven Design
+
+The codebase and project attempts to follow the principles of Domain Driven Design (DDD) as much as possible, and
+wherever it makes sense. DDD is a software development approach that focuses on the domain, the problem that the
+software is trying to solve, and the business logic that drives the problem.
+
+It is a way of thinking and a set of priorities, aimed at accelerating software projects that have to deal with
+complex domains that are constantly changing.
+
+- [Principles of DDD]
```
Comment (hanlun0804) [inline docs/domain-driven-design.md:10]: Should these be links to the respective headers?

**Comment by hanlun0804** [comment]
I think it was a nice introduction, some yapping is nice when you go from not knowing what it is :)

File: docs/domain-driven-design.md (hunk 1 above)
Comment (hanlun0804) [reply docs/domain-driven-design.md:10]: In that case:
- [Principles of DDD](#principles-of-ddd)
- [Hexagonal Architecture](#hexagonal-architecture)
- [Implemented in code](#implemented-in-code)

**Comment by henrikskog** [comment]
Looking good! I would like a lot more examples, but this is something we can iterate on going forward

**Comment by henrikskog** [comment]
It would be nice to include some examples of the patterns we follow in repositories and services. For example, when a `delete` method in a repository returns `null`, it means that there was no such resource to delete. This sort of stuff is important to document I think

**Comment by junlarsen** [comment]
I agree we should add stuff about "return null" and other paradigms we've used, but I think they are probably better fit in another document

File: docs/domain-driven-design.md (hunk 1 above)
Comment (junlarsen) [reply docs/domain-driven-design.md:10]: Yep my bad

**Pull request merged**